src/resolve/resolved-dns-packet.c

```diff
@@ -2221,14 +2221,14 @@ static int dns_packet_extract_question(DnsPacket *p, DnsQuestion **ret_question)
 
                 for (i = 0; i < n; i++) {
                         _cleanup_(dns_resource_key_unrefp) DnsResourceKey *key = NULL;
-                        bool cache_flush;
+                        bool qu;
 
-                        r = dns_packet_read_key(p, &key, &cache_flush, NULL);
+                        r = dns_packet_read_key(p, &key, &qu, NULL);
                         if (r < 0)
                                 return r;
 
-                        if (cache_flush)
-                                return -EBADMSG;
+                        if (qu)
+                                p->wants_unicast_reply = true;
```

Comment:
hmm, this is a bit sloppy, no? mdns allows coalescing multiple questions into one packet, but only set the UP bit for some of these questions. Here you propagate this up: the flag is suddenly applied to all questions of the packet, no?

I figure the right fix for this is probably what we did for DnsAnswer + DnsAnswerItem, where we can store an extra flags field for each item. 

hence I think we should add DnsQuestionItem which is to DnsQuestion what DnsAnswerItem is to DnsAnswer, i.e.:

```c
typedef struct DnsQuestion DnsQuestion;
typedef struct DnsQuestionItem DnsQuestionItem;

typedef DnsQuestionFlags {
        DNS_QUESTION_UNICAST_REQUEST =  1 << 0,
DnsQuestionFlags;

struct DnsQuestionItem {
        DnsResourceKey *key;
        DnsQuestionFlags flags;
};

struct DnsQuestion {
        unsigned n_ref;
        size_t n_keys, n_allocated;
        DnsQuestionItem* items[0];
};
```

Sure right now we'll only define a single flag, but it's nice to have this somewhat symmetric to DnsAnswerFlags, and we might need more sooner or late.

Does this make sense?

Reply:
Good point (how could I miss this!) — although coalescing QU and QM questions into a single packet sounds like a weird thing to do, the RFC is pretty clear that this is allowed & we should just treat them as independent questions.

And yes, storing this per question item would make more sense. I was trying to make the changes here as uninvasive as possible (because this is a rather niche feature), but if you don't mind, sure, I'll try adding `DnsQuestionItem`s.

Reply:
There's also a question of what we should actually do with such mixed queries. Should we reply with two separate packets, one multicast and one unicast? Or, since QU bit is merely a suggestion (unlike the other conditions that cause us to reply over unicast, such as legacy queries) perhaps we should just send the whole reply over multicast in that case?

Reply:
Went for the latter.